From bd586671a3c22eb2f07e55f11b3ce64e1f7961e7 Mon Sep 17 00:00:00 2001
From: Mark Williams <mwilliams@fb.com>
Date: Thu, 20 Feb 2020 15:41:33 -0800
Subject: [PATCH] Fix a buffer-overrun in SimpleParser

Summary: In the failure case, we might have already consumed the entire string.

Reviewed By: binliu19, ottoni

Differential Revision: D19610775

fbshipit-source-id: d387df15994a310f5a31cfbb5fa11679997f7ae7
---
 hphp/runtime/ext/json/JSON_parser.cpp           | 4 ++--
 hphp/test/slow/ext_json/decode_crash.php        | 1 +
 hphp/test/slow/ext_json/decode_crash.php.expect | 1 +
 3 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/hphp/runtime/ext/json/JSON_parser.cpp b/hphp/runtime/ext/json/JSON_parser.cpp
index 5d3b3cfb25a2..b8d52359368d 100644
--- a/hphp/runtime/ext/json/JSON_parser.cpp
+++ b/hphp/runtime/ext/json/JSON_parser.cpp
@@ -342,8 +342,8 @@ struct SimpleParser {
                        JSONContainerType container_type, bool is_tsimplejson) {
     SimpleParser parser(inp, length, buf, container_type, is_tsimplejson);
     bool ok = parser.parseValue();
-    parser.skipSpace();
-    if (!ok || parser.p != inp + length) {
+    if (!ok ||
+        (parser.skipSpace(), parser.p != inp + length)) {
       // Unsupported, malformed, or trailing garbage. Release entire stack.
       tvDecRefRange(buf, parser.top);
       return false;
diff --git a/hphp/test/slow/ext_json/decode_crash.php b/hphp/test/slow/ext_json/decode_crash.php
index 9944145e454e..003b886b2f42 100644
--- a/hphp/test/slow/ext_json/decode_crash.php
+++ b/hphp/test/slow/ext_json/decode_crash.php
@@ -1,3 +1,4 @@
 <?hh
 
 var_dump(json_decode('"a"', false, 0, 0));
+var_dump(json_decode('"abc', true, 1000, 0));
diff --git a/hphp/test/slow/ext_json/decode_crash.php.expect b/hphp/test/slow/ext_json/decode_crash.php.expect
index 7951defec192..e2a4ea7d26ca 100644
--- a/hphp/test/slow/ext_json/decode_crash.php.expect
+++ b/hphp/test/slow/ext_json/decode_crash.php.expect
@@ -1 +1,2 @@
 NULL
+NULL
